---
title: "class11"
author: "Joshua Martin (PID: A18545389)"
format: pdf
---

## AlphaFold Data Base (AFDB)

The EBI maintains the largest database of AlphaFold structure prediction models at: https://alphafold.ebi.ac.uk

From last class, Class09, (before Halloween) we saw that the PDB had 244,290 (Oct 2025)

The total number of protein sequences in UniProtKB is 199,579,901

> **Key Point**: This is a tiny fraction of sequence space that has structural coverage (0.12%)
```{r}
244290/199579901 * 100
```

AFDB is attempting to address this gap...

There are two "Quality Scores" from AlphaFold one for residues (i.e each amino acid) called "pLDDT" socre. The other "PAE" score that measures the confidence in the relative position of two residues (i.e. a score for every pair of residues)

## Generating your own structure predictions

image of all 5 models

```{r}
knitr::include_graphics("5models.png")
```

```{r}
knitr::include_graphics("fin5modelsuperimposed.png")
```



# Section 8

# Custom analysis of resulting models in R

Read key result files into R. The first thing I need to know is what my results directory/folder is called (i.e. its name is different for every AlphaFold run/job)

```{r}
results.dir <- "HIVPR_dimer_23119"

# Change this for YOUR results dir name
results_dir <- "HIVPR_dimer_23119/" 

# File names for all PDB models
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)

# Print our PDB file names
basename(pdb_files)

```

```{r}
library(bio3d)

m1 <- read.pdb(pdb_files[1])
m1
```

```{r}
plot.bio3d(m1$atom$b[m1$calpha])
```

```{r}
head(m1$atom)
```


```{r}
m1$atom$b[m1$calpha]
```

```{r}
plot( m1$atom$b[m1$calpha], typ="l")
```

# Residue conservation from alignment file

Find the large AlphaFold alignment file
```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```

Read this int oR

```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)

```

How many sequences are in this alignment
```{r}
dim(aln$ali)
```

We can score residue conservation in the alignment with the conserv() function.

```{r}
sim <- conserv(aln)

```

```{r}
plotb3(sim[1:99], ylab="Conservation Score")
```

Note the conserved Active Site residues D25, T26, G27, A28. These positions will stand out if we generate a consensus sequence with a high cutoff value:


```{r}
con <- consensus(aln, cutoff = 0.9)
con$seq
```
```{r}
m1.pdb <- read.pdb(pdb_files[1])
occ <- vec2resno(c(sim[1:99], sim[1:99]), m1.pdb$atom$resno)
write.pdb(m1.pdb, o=occ, file="m1_conserv.pdb")
```

